packages/@lwc/engine/src/framework/template.ts

```diff
@@ -175,19 +176,33 @@ export function evaluateTemplate(vm: VM, html: Template): Array<VNode | null> {
     // right before producing the vnodes, we clear up all internal references
     // to custom elements from the template.
     vm.velements = [];
-    // invoke the selected template.
-    const vnodes: VNodes = html.call(undefined, api, component, cmpSlots, context.tplCache!);
 
-    const { styleVNode } = context;
-    if (!isNull(styleVNode)) {
-        ArrayUnshift.call(vnodes, styleVNode);
-    }
-
-    if (process.env.NODE_ENV !== 'production') {
-        assert.invariant(
-            isArray(vnodes),
-            `Compiler should produce html functions that always return an array.`
-        );
-    }
+    const isEvaluatingTemplateInception = isEvaluatingTemplate;
+    let vnodes: VNodes = [];
+    runWithBoundaryProtection(
+        vm,
+        vm.owner,
+        () => {
+            isEvaluatingTemplate = true;
+        },
+        () => {
+            // invoke the selected template.
+            vnodes = html.call(undefined, api, component, cmpSlots, context.tplCache!);
+            const { styleVNode } = context;
+            if (!isNull(styleVNode)) {
+                ArrayUnshift.call(vnodes, styleVNode);
+            }
+
+            if (process.env.NODE_ENV !== 'production') {
+                assert.invariant(
```

Comment:
imagine that this throws an error... that will be captured by the boundary... that's not what we want.